File: clear/src/lexer.cpp
#include "lexer.hpp"

#include <cctype>
#include <iostream>

// Definition of the function
std::string tokenTypeToString(TokenType type) {
  switch (type) {
    case TokenType::LEFT_PAREN:
      return "LEFT_PAREN";
    case TokenType::RIGHT_PAREN:
      return "RIGHT_PAREN";
    case TokenType::LEFT_BRACE:
      return "LEFT_BRACE";
    case TokenType::RIGHT_BRACE:
      return "RIGHT_BRACE";
    case TokenType::LEFT_BRACKET:
      return "LEFT_BRACKET";
    case TokenType::RIGHT_BRACKET:
      return "RIGHT_BRACKET";
    case TokenType::SEMI:
      return "SEMI";
    case TokenType::COMMA:
      return "COMMA";
    case TokenType::DOT:
      return "DOT";
    case TokenType::SINGLE_QUOTE:
      return "SINGLE_QUOTE";
    case TokenType::DOUBLE_QUOTE:
      return "DOUBLE_QUOTE";
    case TokenType::COLON:
      return "COLON";
    case TokenType::BACKSLASH:
      return "BACKSLASH";
    case TokenType::PLUS:
      return "PLUS";
    case TokenType::MINUS:
      return "MINUS";
    case TokenType::STAR:
      return "STAR";
    case TokenType::SLASH:
      return "SLASH";
    case TokenType::PERCENT:
      return "PERCENT";
    case TokenType::CARET:
      return "CARET";
    case TokenType::QUESTION:
      return "QUESTION";
    case TokenType::EQUAL:
      return "EQUAL";
    case TokenType::LESS:
      return "LESS";
    case TokenType::GREATER:
      return "GREATER";
    case TokenType::BANG:
      return "BANG";
    case TokenType::PLUS_EQUAL:
      return "PLUS_EQUAL";
    case TokenType::MINUS_EQUAL:
      return "MINUS_EQUAL";
    case TokenType::STAR_EQUAL:
      return "STAR_EQUAL";
    case TokenType::SLASH_EQUAL:
      return "SLASH_EQUAL";
    case TokenType::EQUAL_EQUAL:
      return "EQUAL_EQUAL";
    case TokenType::LESS_EQUAL:
      return "LESS_EQUAL";
    case TokenType::GREATER_EQUAL:
      return "GREATER_EQUAL";
    case TokenType::BANG_EQUAL:
      return "BANG_EQUAL";
    case TokenType::AND:
      return "AND";
    case TokenType::OR:
      return "OR";
    case TokenType::NUMBER:
      return "NUMBER";
    case TokenType::IDENT:
      return "IDENT";
    case TokenType::LET:
      return "LET";
    case TokenType::RETURN:
      return "RETURN";
    case TokenType::CONST:
      return "CONST";
    case TokenType::IF:
      return "IF";
    case TokenType::ELSE:
      return "ELSE";
    case TokenType::STRING:
      return "STRING";
    case TokenType::BOOL:
      return "BOOL";
    case TokenType::WHILE:
      return "WHILE";
    case TokenType::FOR:
      return "FOR";
    case TokenType::BREAK:
      return "BREAK";
    case TokenType::CONTINUE:
      return "CONTINUE";
    case TokenType::_EOF:
      return "END_OF_FILE";
    default:
      return "UNDEFINED";
  }
}

// Declaration of the keywordLookup function
// Token keywordLookup(const std::string& literal, int line, int col);

char Lexer::peek() const {
  if (pos >= src.size()) {
    return '\0';
  }
  return src[pos];
}

char Lexer::peekN(size_t n) const {
  if (pos + n >= src.size()) {
    return '\0';
  }
  return src[pos + n];
}

char Lexer::consume() {
  if (pos >= src.size()) {
    return '\0';
  }
  char c = src[pos++];
  return c;
}

void Lexer::skipWhitespace() {
  while (peek() == ' ' || peek() == '\t' || peek() == '\n' || peek() == '\r') {
    char c = consume();
    if (c == '\n' || c == '\r') {
      line++;
      column = 1;
    } else if (c == '\t') {
      column += 4;
    } else if (c == ' ') {
      column++;
    }
  }
}

// Definition of the keywordLookup function
Token keywordLookup(const std::string& literal, int line, int col) {
  if (literal == "let") {
    return Token(TokenType::LET, literal, line, col);
  } else if (literal == "return") {
    return Token(TokenType::RETURN, literal, line, col);
  } else if (literal == "const") {
    return Token(TokenType::CONST, literal, line, col);
  } else if (literal == "if") {
    return Token(TokenType::IF, literal, line, col);
  } else if (literal == "else") {
    return Token(TokenType::ELSE, literal, line, col);
  } else if (literal == "string") {
    return Token(TokenType::STRING, literal, line, col);
  } else if (literal == "bool") {
    return Token(TokenType::BOOL, literal, line, col);
  } else if (literal == "while") {
    return Token(TokenType::WHILE, literal, line, col);
  } else if (literal == "for") {
    return Token(TokenType::FOR, literal, line, col);
  } else if (literal == "break") {
    return Token(TokenType::BREAK, literal, line, col);
  } else if (literal == "continue") {
    return Token(TokenType::CONTINUE, literal, line, col);
  } else {
    return Token(TokenType::IDENT, literal, line, col);
  }
}

Token Lexer::nextToken() {
  skipWhitespace();

  char c = peek();

  if (c == '\0') {
    return Token(TokenType::_EOF, "", line, column + 1);
  }

  if (isalpha(c)) {
    std::string literal;
    int startColumn = column;
    while (isalnum(peek()) || peek() == '_') {
      literal += consume();
      column++;
    }
    const Token token = keywordLookup(literal, line, startColumn);
    return token;
  }

  if (isdigit(c)) {
    std::string literal;
    bool isFloat = false;
    int startColumn = column;
    literal += consume();
    column++;

    while (isdigit(peek())) {
      literal += consume();
      column++;
    }

    if (peek() == '.' && isdigit(peekN(1))) {
      literal += consume();
      column++;
      while (isdigit(peek())) {
        literal += consume();
        column++;
      }
    }

    return Token(TokenType::NUMBER, literal, line, startColumn);
  }

  switch (c) {
    case '(':
      consume();
      return Token(TokenType::LEFT_PAREN, "(", line, column);
    case ')':
      consume();
      return Token(TokenType::RIGHT_PAREN, ")", line, column);
    case '{':
      consume();
      return Token(TokenType::LEFT_BRACE, "{", line, column);
    case '}':
      consume();
      return Token(TokenType::RIGHT_BRACE, "}", line, column);
    case '[':
      consume();
      return Token(TokenType::LEFT_BRACKET, "[", line, column);
    case ']':
      consume();
      return Token(TokenType::RIGHT_BRACKET, "]", line, column);
    case ';':
      consume();
      return Token(TokenType::SEMI, ";", line, column);
    case ',':
      consume();
      return Token(TokenType::COMMA, ",", line, column);
    case '.':
      consume();
      return Token(TokenType::DOT, ".", line, column);
    case '\'':
      consume();
      return Token(TokenType::SINGLE_QUOTE, "'", line, column);
    case '"':
      consume();
      return Token(TokenType::DOUBLE_QUOTE, "\"", line, column);
    case ':':
      consume();
      return Token(TokenType::COLON, ":", line, column);
    case '\\':
      consume();
      return Token(TokenType::BACKSLASH, "\\", line, column);
    case '+':
      if (peekN(1) == '=') {
        consume();
        consume();
        return Token(TokenType::PLUS_EQUAL, "+=", line, column);
      }
      consume();
      return Token(TokenType::PLUS, "+", line, column);
    case '-':
      if (peekN(1) == '=') {
        consume();
        consume();
        return Token(TokenType::MINUS_EQUAL, "-=", line, column);
      }
      consume();
      return Token(TokenType::MINUS, "-", line, column);
    case '*':
      if (peekN(1) == '=') {
        consume();
        consume();
        return Token(TokenType::STAR_EQUAL, "*=", line, column);
      }
      consume();
      return Token(TokenType::STAR, "*", line, column);
    case '/':
      if (peekN(1) == '=') {
        consume();
        consume();
        return Token(TokenType::SLASH_EQUAL, "/=", line, column);
      }
      consume();
      return Token(TokenType::SLASH, "/", line, column);
    case '%':
      consume();
      return Token(TokenType::PERCENT, "%", line, column);
    case '^':
      consume();
      return Token(TokenType::CARET, "^", line, column);
    case '?':
      consume();
      return Token(TokenType::QUESTION, "?", line, column);
    case '=':
      if (peekN(1) == '=') {
        consume();
        consume();
        return Token(TokenType::EQUAL_EQUAL, "==", line, column);
      }
      consume();
      return Token(TokenType::EQUAL, "=", line, column);
    case '<':
      if (peekN(1) == '=') {
        consume();
        consume();
        return Token(TokenType::LESS_EQUAL, "<=", line, column);
      }
      consume();
      return Token(TokenType::LESS, "<", line, column);
    case '>':
      if (peekN(1) == '=') {
        consume();
        consume();
        return Token(TokenType::GREATER_EQUAL, ">=", line, column);
      }
      consume();
      return Token(TokenType::GREATER, ">", line, column);
    default:
      consume();
      return Token(TokenType::UNDEFINED, std::string(1, c), line, column);
  }
}

std::vector<Token> Lexer::tokenize() {
  std::vector<Token> tokens;
  Token token = nextToken();
  while (token.getType() != TokenType::_EOF) {
    tokens.push_back(token);
    token = nextToken();
  }
  tokens.push_back(token);  // EOF token
  return tokens;
}

File: goclear/lexer/lexer.go
package lexer

import "github.com/ajtroup1/goclear/token"

var keywords = map[string]token.TokenType{
	"true":     token.BOOL,
	"false":    token.BOOL,
	"if":       token.IF,
	"else":     token.ELSE,
	"for":      token.FOR,
	"while":    token.WHILE,
	"fn":       token.FUNCTION,
	"let":      token.LET,
	"return":   token.RETURN,
	"break":    token.BREAK,
	"continue": token.CONTINUE,
	"null":     token.NULL,
	"new":      token.NEW,
	"class":    token.CLASS,
	"this":     token.THIS,
	"super":    token.SUPER,
	"static":   token.STATIC,
	"import":   token.IMPORT,
	"export":   token.EXPORT,
}

type Lexer struct {
	src     string
	pos     int
	readPos int
	c       byte

	line int
	col  int
}

func New(src string) *Lexer {
	l := &Lexer{
		src:  src,
		line: 1,
		col:  0,
	}
	l.readChar()
	return l
}

func (l *Lexer) Lex() []token.Token {
	var tokens []token.Token
	for l.c != 0 {
		tok := l.nextToken()
		tokens = append(tokens, tok)
		if tok.Type == token.EOF {
			break
		}
	}
	return tokens
}

func (l *Lexer) nextToken() token.Token {
	l.skipWhitespace()

	var tok token.Token

	switch l.c {
	case 0:
		tok = token.Token{Type: token.EOF, Literal: "", Line: l.line, Col: l.col}
	case '=':
		if l.peek() == '=' {
			l.readChar()
			tok = token.Token{Type: token.EQ, Literal: "==", Line: l.line, Col: l.col}
		} else {
			tok = token.Token{Type: token.ASSIGN, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	case '+':
		if l.peek() == '=' {
			l.readChar()
			tok = token.Token{Type: token.PLUS_EQ, Literal: "+=", Line: l.line, Col: l.col}
		} else if l.peek() == '+' {
			l.readChar()
			tok = token.Token{Type: token.INC, Literal: "++", Line: l.line, Col: l.col}
		} else {
			tok = token.Token{Type: token.PLUS, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	case '-':
		if l.peek() == '=' {
			l.readChar()
			tok = token.Token{Type: token.MINUS_EQ, Literal: "-=", Line: l.line, Col: l.col}
		} else if l.peek() == '-' {
			l.readChar()
			tok = token.Token{Type: token.DEC, Literal: "--", Line: l.line, Col: l.col}
		} else {
			tok = token.Token{Type: token.MINUS, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	case '!':
		if l.peek() == '=' {
			l.readChar()
			tok = token.Token{Type: token.NOT_EQ, Literal: "!=", Line: l.line, Col: l.col}
		} else {
			tok = token.Token{Type: token.BANG, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	case '*':
		if l.peek() == '=' {
			l.readChar()
			tok = token.Token{Type: token.MUL_EQ, Literal: "*=", Line: l.line, Col: l.col}
		} else {
			tok = token.Token{Type: token.ASTERISK, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	case '/':
		if l.peek() == '=' {
			l.readChar()
			tok = token.Token{Type: token.DIV_EQ, Literal: "/=", Line: l.line, Col: l.col}
		} else {
			tok = token.Token{Type: token.SLASH, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	case '<':
		if l.peek() == '=' {
			l.readChar()
			tok = token.Token{Type: token.LT_EQ, Literal: "<=", Line: l.line, Col: l.col}
		} else {
			tok = token.Token{Type: token.LT, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	case '>':
		if l.peek() == '=' {
			l.readChar()
			tok = token.Token{Type: token.GT_EQ, Literal: ">=", Line: l.line, Col: l.col}
		} else {
			tok = token.Token{Type: token.GT, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	case ';':
		tok = token.Token{Type: token.SEMICOLON, Literal: string(l.c), Line: l.line, Col: l.col}
	case ',':
		tok = token.Token{Type: token.COMMA, Literal: string(l.c), Line: l.line, Col: l.col}
	case ':':
		tok = token.Token{Type: token.COLON, Literal: string(l.c), Line: l.line, Col: l.col}
	case '(':
		tok = token.Token{Type: token.LPAREN, Literal: string(l.c), Line: l.line, Col: l.col}
	case ')':
		tok = token.Token{Type: token.RPAREN, Literal: string(l.c), Line: l.line, Col: l.col}
	case '{':
		tok = token.Token{Type: token.LBRACE, Literal: string(l.c), Line: l.line, Col: l.col}
	case '}':
		tok = token.Token{Type: token.RBRACE, Literal: string(l.c), Line: l.line, Col: l.col}
	case '[':
		tok = token.Token{Type: token.LBRACKET, Literal: string(l.c), Line: l.line, Col: l.col}
	case ']':
		tok = token.Token{Type: token.RBRACKET, Literal: string(l.c), Line: l.line, Col: l.col}
	default:
		if isLetter(l.c) {
			return l.readIdentifier()
		} else if isDigit(l.c) {
			return l.readNumber()
		} else {
			tok = token.Token{Type: token.ILLEGAL, Literal: string(l.c), Line: l.line, Col: l.col}
		}
	}

	l.readChar()
	return tok
}

func (l *Lexer) readIdentifier() token.Token {
	startLine, startCol := l.line, l.col

	pos := l.pos
	for isLetter(l.c) {
		l.readChar()
	}
	literal := l.src[pos:l.pos]
	return keywordLookup(literal, startLine, startCol)
}

func keywordLookup(ident string, line, col int) token.Token {
	if tok, found := keywords[ident]; found {
		return token.Token{Type: tok, Literal: ident, Line: line, Col: col}
	}
	return token.Token{Type: token.IDENT, Literal: ident, Line: line, Col: col}
}

func (l *Lexer) readNumber() token.Token {
	startLine, startCol := l.line, l.col

	pos := l.pos
	isFloat := false

	for isDigit(l.c) || l.c == '.' {
		if l.c == '.' {
			isFloat = true
		}
		l.readChar()
	}

	literal := l.src[pos:l.pos]

	if isFloat {
		return token.Token{Type: token.FLOAT, Literal: literal, Line: startLine, Col: startCol}
	}
	return token.Token{Type: token.INT, Literal: literal, Line: startLine, Col: startCol}
}

func (l *Lexer) skipWhitespace() {
	for isWhitespace(l.c) {
		l.readChar()
	}
}

func (l *Lexer) readChar() {
	if l.readPos >= len(l.src) {
		l.c = 0
	} else {
		l.c = l.src[l.readPos]
	}

	if l.c == '\n' || l.c == '\r' {
		l.line++
		l.col = 0
	} else {
		l.col++
	}

	l.pos = l.readPos
	l.readPos++
}

func (l *Lexer) peek() byte {
	if l.readPos >= len(l.src) {
		return 0
	}

	return l.src[l.readPos]
}

func (l *Lexer) peekN(n int) byte {
	if l.readPos+n >= len(l.src) {
		return 0
	}

	return l.src[l.readPos+n]
}

func (l *Lexer) consume() {
	if l.readPos >= len(l.src) {
		l.c = 0
	} else {
		l.c = l.src[l.readPos]
	}
}

func isWhitespace(c byte) bool {
	return c == ' ' || c == '\t' || c == '\n' || c == '\r'
}
func isLetter(c byte) bool {
	return 'a' <= c && c <= 'z' || 'A' <= c && c <= 'Z' || c == '_'
}

func isDigit(c byte) bool {
	return '0' <= c && c <= '9'
}
